### PR TITLE
applications: nrf_desktop: DFU state checking

### PR DESCRIPTION
**Pull request by pdunaj**
Jira:DESK-574

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

File: scripts/hid_configurator/configurator.py [hunk 1]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
     img_length = os.stat(dfu_image).st_size
     print('DFU image size: {} bytes'.format(img_length))
 
+    for i in range(3):
+        dfu_info = dfu_sync(dev, args)
+        if dfu_info is not None:
+            if dfu_info[0] != 0:
+                time.sleep(1)
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator.py]: Maybe add comment to explain why we need sleep here

File: scripts/hid_configurator/configurator.py [hunk 2]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
...
-    finally:
-        img_file.close()
+
+    img_file.close()
+    print('')
 
     if success:
-        print('')
-        print('DFU transfer completed')
+        for i in range(3):
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator.py]: Name the constant

File: scripts/hid_configurator/configurator.py [hunk 3]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
...
+
+    if dfu_info[0] != 0:
+        print('Cannot start DFU, already in progress')
+        return
+
+    img_csum = file_crc(dfu_image)
+    if (dfu_info[1] == img_length) and (dfu_info[2] == img_csum) and (dfu_info[3] <= img_length):
+        print('Resume DFU at {}'.format(dfu_info[3]))
+        offset = dfu_info[3]
+    else:
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator.py]: Add print to indicate that we start DFU ?

File: scripts/hid_configurator/configurator.py [hunk 4]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
...
+        offset = 0
+
+    dfu_start(dev, args, img_length, img_csum, offset)
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator.py]: Error checking?

File: scripts/hid_configurator/configurator.py [hunk 5]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
...
-        while True:
-            dfu_header = struct.pack('<II', img_length, offset)
-
-            chunk_data = img_file.read(EVENT_DATA_LEN_MAX - len(dfu_header))
+        while offset < img_length:
+            if offset % 4096 == 0:
+                success = False
+                dfu_info = dfu_sync(dev, args)
+                if dfu_info is None:
+                    break
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator.py]: Add print informing about failure reason (e.g. device not responding)

File: scripts/hid_configurator/configurator.py [hunk 6]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
...
+    print('')
 
     if success:
-        print('')
-        print('DFU transfer completed')
+        for i in range(3):
+            dfu_info = dfu_sync(dev, args)
+            if dfu_info is not None:
+                if dfu_info[0] != 0:
+                    time.sleep(1)
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator.py]: Maybe comment why we need to sleep here?

File: scripts/hid_configurator/configurator.py [hunk 7]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
...
-    offset = 0
+    img_file.seek(offset)
 
     try:
-        while True:
-            dfu_header = struct.pack('<II', img_length, offset)
-
-            chunk_data = img_file.read(EVENT_DATA_LEN_MAX - len(dfu_header))
+        while offset < img_length:
+            if offset % 4096 == 0:
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator.py]: We may think about adding name for 4096

File: applications/nrf_desktop/src/modules/dfu.c [hunk 8]
```diff
@@ -204,6 +159,116 @@ static void handle_dfu_data(const struct config_event *event)
...
+	}
+
+	k_delayed_work_submit(&dfu_timeout, DFU_TIMEOUT);
+}
+
+static void handle_dfu_sync(const struct config_fetch_request_event *event)
+{
+	LOG_INF("DFU sync requested");
+
+	u8_t dfu_active = (flash_area != NULL) ? 0x01 : 0x00;
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/dfu.c]: I prefer to leave it as is as when it is done this way I see a byte being written. The same goes for writing 0x0001 for 16-bit and 0x00000001 for 32-bit constant data.

File: scripts/hid_configurator/configurator.py (hunk 4 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: error is checked by dfu_sync. if this fails dfu will not be started and dfu_sync inform us about it
if there is a communication error dfu_sync inform about it

File: scripts/hid_configurator/configurator.py [hunk 9]
```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
     img_length = os.stat(dfu_image).st_size
     print('DFU image size: {} bytes'.format(img_length))
 
+    for i in range(3):
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Done

File: scripts/hid_configurator/configurator.py (hunk 1 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Done

File: scripts/hid_configurator/configurator.py (hunk 3 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: I don't think it is needed as progress bar indicates that

File: scripts/hid_configurator/configurator.py (hunk 4 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Added

File: scripts/hid_configurator/configurator.py (hunk 7 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Done

File: scripts/hid_configurator/configurator.py (hunk 5 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Done

File: scripts/hid_configurator/configurator.py (hunk 2 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Done

File: scripts/hid_configurator/configurator.py (hunk 6 above)
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Done

File: applications/nrf_desktop/src/modules/dfu.c [hunk 10]
```diff
@@ -204,6 +159,116 @@ static void handle_dfu_data(const struct config_event *event)
...
+		return;
+	}
+
+	size_t pos = 0;
+
+	length = sys_get_le32(&data[pos]);
+	pos += sizeof(img_length);
+
+	csum = sys_get_le32(&data[pos]);
+	pos += sizeof(img_csum);
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/dfu.c]: Done

File: scripts/hid_configurator/configurator.py [hunk 11]
```diff
@@ -251,6 +254,59 @@ def open_device(device_type):
...
+def dfu_sync(dev, args):
+    event_id = (EVENT_GROUP_DFU << GROUP_FIELD_POS) | (DFU_SYNC << TYPE_FIELD_POS)
+    recipient = get_device_pid(args.device_type)
+
+    try:
+        success, fetched_data = exchange_feature_report(dev, recipient, event_id, None, True)
+    except:
+        success = False
+
+    fmt = '<BIII'
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Fixed

File: scripts/hid_configurator/configurator.py [hunk 12]
```diff
@@ -251,6 +254,59 @@ def open_device(device_type):
...
+    assert(struct.calcsize(fmt) <= EVENT_DATA_LEN_MAX)
+
+    if not success:
+        return
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply scripts/hid_configurator/configurator.py]: Fixed

File: applications/nrf_desktop/src/modules/dfu.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/dfu.c]: ok

**Comment by pdunaj** [comment]
Hi @wiminordic , can you give a green tick?

**Pull request merged**